core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts

```diff
@@ -0,0 +1,304 @@
+        AppConstants.CONTRIBUTION_STATS_SUBTYPE_REVIEW] =
+          this.getObtainedBadges(
+            this.totalTranslationStats[language].reviews,
+            AppConstants.CONTRIBUTION_STATS_SUBTYPE_REVIEW,
+            this.totalTranslationStats[language].language);
+      this.translationBadges[language][
+        AppConstants.CONTRIBUTION_STATS_SUBTYPE_CORRECTION] =
+          this.getObtainedBadges(
+            this.totalTranslationStats[language].corrections,
+            AppConstants.CONTRIBUTION_STATS_SUBTYPE_CORRECTION,
+            this.totalTranslationStats[language].language);
+    }
+
+    this.questionSubmissionBadges = this.getObtainedBadges(
+      this.totalQuestionStats.submissions,
+      AppConstants.CONTRIBUTION_STATS_SUBTYPE_SUBMISSION);
+    this.questionReviewBadges = this.getObtainedBadges(
+      this.totalQuestionStats.reviews,
+      AppConstants.CONTRIBUTION_STATS_SUBTYPE_REVIEW);
+    this.questionCorrectionBadges = this.getObtainedBadges(
+      this.totalQuestionStats.corrections,
+      AppConstants.CONTRIBUTION_STATS_SUBTYPE_CORRECTION);
+
+    this.languages = Object.keys(this.totalTranslationStats);
+
+    if (this.languages.length > 0) {
+      this.selectedLanguage = this.languages[0];
+      this.userCanReviewTranslationSuggestion = (
+        this.reviewableLanguages.includes(this.selectedLanguage));
+    }
+    this.dataLoading = false;
+  }
+
+  getObtainedBadges(
+      contributionCount: number,
+      contributionSubType: string, language?: string): Badge[] {
+    const badges: Badge[] = [];
+    for (const level of AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS) {
+      console.log(contributionCount);
+      if (contributionCount >= level) {
+        badges.push(
+          {
+            contributionCount: level,
+            text: contributionSubType,
+            isUnlocked: true,
+            language
+          }
+        );
+      } else {
+        // From this block, we are adding locked badge(in predefined range)
+        // which will be achieved next.
+        badges.push(
+          {
+            contributionCount: level - contributionCount,
+            text: contributionSubType,
+            isUnlocked: false,
+            language
+          }
+        );
+        return badges;
+      }
+    }
+
+    let level = 500;
+    while (contributionCount >= level) {
+      level += 500;
+      if (contributionCount >= level) {
+        badges.push(
+          {
+            contributionCount: level,
+            text: contributionSubType,
+            isUnlocked: true,
+            language
+          }
+        );
+      } else {
+        // From this block, we are adding locked badge(in non-predefined range)
+        // which will be achieved next.
+        badges.push(
+          {
+            contributionCount: level - contributionCount,
+            text: contributionSubType,
+            isUnlocked: false,
+            language
+          }
+        );
+        return badges;
```

Comment:
Since this logic is the same as above, let's consolidate this into one loop (keep track of index, increment index, when index gets to end of `AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS`, start incrementing level by 500 instead).

Reply:
Done